### PR TITLE
torch.linalg.det

### LINKED ISSUE #51652: Move torch.det implementation to linalg namespace and add out= argument.
Currently, torch.linalg.det is an alias for torch.det. We should move the implementation of torch.det to the linalg namespace and also add an out= parameter. Then make torch.det alias torch.linalg.det instead

cc @jianyuh @nikitaved @pearu @mruberry @heitorschueroff @walterddr @IvanYashchuk

### PR DESCRIPTION
**Pull request by antocuni**
Fixes #51652.
In particular:
- the main implementation is in `torch.linalg.det` now. `torch.det` is just a deprecated alias to it
- add a new `OpInfo` for `torch.linalg.det`
- remove the old-style tests for `torch.det` (this is similar to what we did for `torch.linalg.slogdet`, see #49194)
- added a `out=` argument to `torch.linalg.det`, but **not** to `torch.det`.

It is worth noting that I had to skip few tests:
- `TestGradientsCuda::test_fn_gradgrad_linalg_det_cuda_float64`. This is not a regression: the functionality is broken also on master, but the test is not executed properly due to #53361.

And the following tests which fails only on ROCm:
- `test_variant_consistency_jit_cuda_{float64,float32}`
- `test_fn_grad_cuda_float64`

I think that the ROCm tests fail because the current linalg.det backward is unstable if the matrix has repeated singular values, see #53364 .

(At the moment of writing some CI jobs are still running but I believe the build will be green, since the only difference wrt the last push is the skip of the ROCm tests)

### CODE REVIEW

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -45,22 +45,34 @@ static inline std::tuple<Tensor, Tensor> _lu_det_P_diag_U(const Tensor& self) {
...
-Tensor linalg_det(const Tensor& self) {
-  return self.det();
+// torch.det, alias for torch.linalg.det
+Tensor det(const Tensor& self) {
+  return at::linalg_det(self);
 }
 
-Tensor det(const Tensor& self) {
+Tensor& linalg_det_out(const Tensor& self, Tensor& out) {
+  checkSameDevice("linalg_det", out, self, "out");
```
Comment (mruberry) [inline aten/src/ATen/native/LinearAlgebra.cpp:54]: Let's make sure to use nice user facing names in error messages like "torch.linalg.det()" instead of "linalg_det"

File: torch/linalg/__init__.py [hunk 2]
```diff
@@ -177,6 +180,12 @@
     >>> torch.linalg.det(a)
     tensor(0.0934)
 
+    >>> out = torch.Tensor()
```
Comment (mruberry) [inline torch/linalg/__init__.py:183]: torch.tensor(), the legacy torch.Tensor() should be considered deprecated

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -2089,6 +2112,28 @@ def _make_tensor_helper(shape, low=None, high=None):
...
+           aten_name='linalg_det',
+           sample_inputs_func=sample_inputs_linalg_det,
+           decorators=[skipCUDAIfNoMagma, skipCPUIfNoLapack],
+           test_complex_grad=False,
+           test_inplace_grad=False,
+           skips=(
+               # https://github.com/pytorch/pytorch/issues/53361
+               SkipInfo('TestGradients', 'test_fn_gradgrad', device_type='cuda',
+                        dtypes=[torch.float64]),
+               # The following tests fail only on ROCm. This is probably
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2127]: The skips can be limited to ROCm only by using `active_if=TEST_WITH_ROCM`

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -2089,6 +2112,28 @@ def _make_tensor_helper(shape, low=None, high=None):
...
+    OpInfo('linalg.det',
+           op=torch.linalg.det,
+           dtypes=floating_and_complex_types(),
+           aten_name='linalg_det',
+           sample_inputs_func=sample_inputs_linalg_det,
+           decorators=[skipCUDAIfNoMagma, skipCPUIfNoLapack],
+           test_complex_grad=False,
+           test_inplace_grad=False,
+           skips=(
+               # https://github.com/pytorch/pytorch/issues/53361
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2124]: Let's file a new issue specifically for torch.linalg.det's autograd failure

File: aten/src/ATen/native/native_functions.yaml [hunk 5]
```diff
@@ -9002,15 +9002,22 @@
...
+  dispatch:
+    DefaultBackend: linalg_det
+
+- func: linalg_det.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
+  python_module: linalg
+  dispatch:
+    DefaultBackend: linalg_det_out
 
+# torch.det, alias for torch.linalg.det
 - func: det(Tensor self) -> Tensor
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:9017]: Does the autograd for det work after these changes? I always get confused about how math and defaultbackend relate in these cases.

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -359,6 +359,29 @@ def sample_inputs_tensor_split(op_info, device, dtype, requires_grad):
...
+        random_square_matrix_of_rank(S, 2, **kw),  # rank2
+        random_fullrank_matrix_distinct_singular_value(S, **kw),  # distinct_singular_value
+        make_tensor((3, 3, S, S), **kw),  # batched
+        make_tensor((3, 3, 1, 1), **kw),  # batched_1x1
+        random_symmetric_matrix(S, 3, **kw),  # batched_symmetric
+        random_symmetric_psd_matrix(S, 3, **kw),  # batched_symmetric_psd
+        random_symmetric_pd_matrix(S, 3, **kw),  # batched_symmetric_pd
+        random_fullrank_matrix_distinct_singular_value(S, 3, 3, **kw),  # batched_distinct_singular_values
+    ]
+    for t in inputs:
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:381]: Can `requires_grad` be added to `kw` above to avoid this bit?

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -2089,6 +2112,28 @@ def _make_tensor_helper(shape, low=None, high=None):
                         # see discussion https://github.com/pytorch/pytorch/pull/47761#issuecomment-747316775
                         SkipInfo('TestGradients', 'test_fn_gradgrad', device_type='cuda'),)
                     ),
+    OpInfo('linalg.det',
+           op=torch.linalg.det,
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2116]: Add 'det' as an alias like this:

https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/testing/_internal/common_methods_invocations.py#L1820

**Review by mruberry — COMMENTED** [review]
Hey @antocuni! This looks good overall. I made a few suggestions and just want to sanity check that autograd works as expected. The math/defaultbackend thing always gets me.

@IvanYashchuk, would you like to take a look at this, too?

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -359,6 +359,29 @@ def sample_inputs_tensor_split(op_info, device, dtype, requires_grad):
                         args=(torch.tensor([1, 2, 3]),),
                         kwargs=dict(dim=1)),)
 
+def sample_inputs_linalg_det(op_info, device, dtype, requires_grad):
+    kw = dict(device=device, dtype=dtype)
+    inputs = [
```
Comment (IvanYashchuk) [inline torch/testing/_internal/common_methods_invocations.py:364]: Does it work for 0x0 and 0xSxS inputs? I think it should be included here at the end of list (putting it first might break "out" tests).

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -2089,6 +2112,28 @@ def _make_tensor_helper(shape, low=None, high=None):
                         # see discussion https://github.com/pytorch/pytorch/pull/47761#issuecomment-747316775
                         SkipInfo('TestGradients', 'test_fn_gradgrad', device_type='cuda'),)
                     ),
+    OpInfo('linalg.det',
+           op=torch.linalg.det,
+           dtypes=floating_and_complex_types(),
+           aten_name='linalg_det',
+           sample_inputs_func=sample_inputs_linalg_det,
+           decorators=[skipCUDAIfNoMagma, skipCPUIfNoLapack],
```
Comment (IvanYashchuk) [inline torch/testing/_internal/common_methods_invocations.py:2120]: Maybe add here `skipCUDAIfRocm` instead of ROCm skips below.

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (IvanYashchuk) [reply torch/testing/_internal/common_methods_invocations.py:2124]: I wrote in the issue that the analysis that the autograd test for det was never run on cuda is not correct https://github.com/pytorch/pytorch/issues/53361#issuecomment-791588410. The reason for this fail https://app.circleci.com/pipelines/github/pytorch/pytorch/281078/workflows/bd828f67-1892-45ba-81cf-164b352ead16/jobs/11315029 most probably is not the error in the implementation. My guess that the problem is not in backward implementation is supported by the fact that only gradgradcheck fails and gradcheck passes.
Looking at the Jacobian values my best guess that the problem is with the numerical error of finite differences to estimate the "numerical Jacobian".

We shouldn't skip it, the test was working before. Can you tell for which SampleInput specifically the test fails?

**Review by IvanYashchuk — COMMENTED** [review]
All looks good!
I only have concerns regarding the skipping of gradgradcheck for CUDA. I think this PR shouldn't add this skip.

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:2124]: Thanks for the analysis @IvanYashchuk, that's really interesting.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 1 above)
Comment (antocuni) [reply aten/src/ATen/native/LinearAlgebra.cpp:59]: good idea, fixed by d29e104a9f

File: torch/linalg/__init__.py (hunk 2 above)
Comment (antocuni) [reply torch/linalg/__init__.py:183]: fixed by dddc37b924.
`torch.tensor()` alone doesn't work, it requires a shape. I ended up using `torch.empty(0)` which looks clearer than `torch.tensor([])` to me.

File: aten/src/ATen/native/native_functions.yaml (hunk 5 above)
Comment (antocuni) [reply aten/src/ATen/native/native_functions.yaml:8689]: yes, that's my understanding as well. Moreover now it has been renamed to `CompositeAutogradImplicit` and it's the default, so I fixed my `native_functions.yml` accordingly

**Comment by ezyang** [comment]
ACKing native_functions.yaml chagnes

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (antocuni) [reply torch/testing/_internal/common_methods_invocations.py:2813]: I disagree with the analysis (or better, I agree with it but it doesn't apply in this specific case).
I wrote up a detailed explanation here: https://github.com/pytorch/pytorch/issues/53361#issuecomment-809493569

TL;DR: `test_gradgrad` does NOT convert inputs to the correct device/dtype. The error that we see here is present also on master, if you put the correct device in the sample inputs.

That said, I also think that the implementation is correct and that the error that we see here is just an artifact produced by gradcheck. I admit I have no idea how to solve though, apart from skipping it.

Also, I don't think that this should block the merge of the PR since it's not a regression. On the contrary, it is an improvement over master, were `torch.det` autograd on cuda is completely untested.

File: torch/testing/_internal/common_methods_invocations.py (hunk 6 above)
Comment (antocuni) [reply torch/testing/_internal/common_methods_invocations.py:406]: not easily because things like `random_symmetric_matrix` & co. don't accept `requires_grad` as a kw arg.
The same pattern is also used by e.g. `sample_inputs_linalg_matrix_power`, so I though it was a good idea to be consistent.

File: torch/testing/_internal/common_methods_invocations.py (hunk 7 above)
Comment (antocuni) [reply torch/testing/_internal/common_methods_invocations.py:2805]: done, 6ff2599c87

File: torch/testing/_internal/common_methods_invocations.py (hunk 9 above)
Comment (antocuni) [reply torch/testing/_internal/common_methods_invocations.py:2809]: we don't want to skip all ROCm tests, only the few which actually fail

File: torch/testing/_internal/common_methods_invocations.py (hunk 8 above)
Comment (antocuni) [reply torch/testing/_internal/common_methods_invocations.py:389]: good idea, I added them in daccdb33d6

**Comment by antocuni** [comment]
@mruberry @IvanYashchuk I think I addressed all the comments you made so it should be ready for another round of review (assuming that there are no unexpected failures, will check that later).

The only remaining bit is to decide what to do with the test skipped because of #53361

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (IvanYashchuk) [reply torch/testing/_internal/common_methods_invocations.py:2816]: The real problem is that for inputs with a rank less than `n-1` the backward is unstable.

<details>
  <summary>Here is the code that demonstrates the failure on CPU.  Setting seed to 4 the tests fail, setting seed to 5 the tests pass.</summary>


```python
import torch
import numpy as np
from torch.testing._internal.common_utils import random_square_matrix_of_rank
from torch.autograd.gradcheck import gradcheck, gradgradcheck

torch.manual_seed(4) # with seed 5 gradgradcheck passes

S = 3
dtype = torch.float64
device = 'cpu'

kw = dict(device=device, dtype=dtype)
inputs = [
    random_square_matrix_of_rank(S, S - 2, **kw),
]

print(inputs)

A = inputs[0].clone()
A.requires_grad_()

def svd_based_det(x):
  _, s, _ = torch.svd(x)
  return torch.prod(s)


gradcheck(torch.linalg.det, A) # this passes
gradgradcheck(torch.linalg.det, A) # this fails

gradcheck(svd_based_det, A) # this passes
gradgradcheck(svd_based_det, A) # this fails
```

</details>

Please comment out dim2_null, rank1, rank2 sample_inputs with a comment pointing to the issue about backward stability of det https://github.com/pytorch/pytorch/issues/53364.
Then this skip of `test_fn_gradgrad` can be removed.

Comment (antocuni) [reply torch/testing/_internal/common_methods_invocations.py]: wow, thank you very much for the analysis. I implemented your suggestions in b71452856f

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 10]
```diff
@@ -50,22 +50,34 @@ static inline std::tuple<Tensor, Tensor> _lu_det_P_diag_U(const Tensor& self) {
...
+  return at::linalg_det(self);
 }
 
-Tensor det(const Tensor& self) {
+Tensor& linalg_det_out(const Tensor& self, Tensor& out) {
+  checkSameDevice("torch.linalg.det", out, self, "out");
+  checkLinalgCompatibleDtype("torch.linalg.det", out, self, "out");
   squareCheckInputs(self);
   TORCH_CHECK((at::isFloatingType(self.scalar_type()) || at::isComplexType(self.scalar_type())),
               "Expected a floating point tensor as input");
```
Comment (mruberry) [inline aten/src/ATen/native/LinearAlgebra.cpp:63]: Nit (not a blocking issue): "Expected a floating point or complex tensor as input."

File: torch/testing/_internal/common_methods_invocations.py [hunk 11]
```diff
@@ -24,8 +24,8 @@
      skipCUDAIfRocm, expectedAlertNondeterministic, precisionOverride,)
 from torch.testing._internal.common_cuda import CUDA11OrLater
 from torch.testing._internal.common_utils import \
-    (random_square_matrix_of_rank,
-     random_symmetric_matrix, random_symmetric_psd_matrix,
+    (random_symmetric_matrix, random_symmetric_psd_matrix,
+     # random_square_matrix_of_rank,
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:28]: This comment can be removed

File: torch/testing/_internal/common_methods_invocations.py [hunk 12]
```diff
@@ -2777,6 +2808,26 @@ def reference_lgamma(x):
...
+           supports_inplace_autograd=False,
+           skips=(
+               # The following tests fail only on ROCm. This is probably
+               # related to the fact that the current linalg.det backward is
+               # unstable if the matrix has repeated singular values, see
+               # https://github.com/pytorch/pytorch/issues/53364
+               SkipInfo('TestGradients', 'test_fn_grad', device_type='cuda',
+                        dtypes=[torch.float64], active_if=TEST_WITH_ROCM),
+               SkipInfo('TestCommon', 'test_variant_consistency_jit', device_type='cuda',
+                        dtypes=[torch.float64, torch.float32], active_if=TEST_WITH_ROCM),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2828]: Nit prefer tuples to lists if the iterable doesn't need to be mutated

**Review by mruberry — APPROVED** [review]
Cool!

@IvanYashchuk, would you like to take a last look?

One follow-up task from this PR is to update the documentation for torch.det() to state it's an alias for torch.linalg.det() and then to docs deprecate it. This can be done during the linalg doc review for the 1.9 release.

**Comment by mruberry** [comment]
Temporarily pulling approval because ONNX failures are real:

```
ar 30 10:22:54 RuntimeError: Exporting the operator linalg_det to ONNX opset version 11 is not supported. Please feel free to request support or submit a pull request on PyTorch GitHub.
```

and ROCm failure is real:

```
18:03:53 ======================================================================
18:03:53 ERROR [20.371s]: test_fn_gradgrad_linalg_det_cuda_float64 (__main__.TestGradientsCUDA)
18:03:53 ----------------------------------------------------------------------
18:03:53 Traceback (most recent call last):
18:03:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 955, in wrapper
18:03:53     method(*args, **kwargs)
18:03:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 297, in instantiated_test
18:03:53     raise rte
18:03:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 292, in instantiated_test
18:03:53     result = test_fn(self, *args)
18:03:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 578, in dep_fn
18:03:53     return fn(slf, device, *args, **kwargs)
18:03:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 578, in dep_fn
18:03:53     return fn(slf, device, *args, **kwargs)
18:03:53   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 266, in test_wrapper
18:03:53     return test(*args, **kwargs)
18:03:53   File "test_ops.py", line 171, in test_fn_gradgrad
18:03:53     self._gradgrad_test_helper(device, dtype, op, op.get_op())
18:03:53   File "test_ops.py", line 139, in _gradgrad_test_helper
18:03:53     return self._check_helper(device, dtype, op, variant, 'gradgradcheck')
18:03:53   File "test_ops.py", line 127, in _check_helper
18:03:53     check_grad_dtypes=True))
18:03:53   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/gradcheck.py", line 722, in gradgradcheck
18:03:53     check_grad_dtypes=check_grad_dtypes, check_batched_grad=check_batched_grad)
18:03:53   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/gradcheck.py", line 612, in gradcheck
18:03:53     return fail_test(get_notallclose_msg(a, n, i, j))
18:03:53   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/gradcheck.py", line 564, in fail_test
18:03:53     raise RuntimeError(msg)
18:03:53 RuntimeError: Jacobian mismatch for output 0 with respect to input 0,
18:03:53 numerical:tensor([[nan, nan, nan,  ..., nan, nan, nan],
18:03:53         [nan, nan, nan,  ..., nan, nan, nan],
18:03:53         [nan, nan, nan,  ..., nan, nan, nan],
18:03:53         ...,
18:03:53         [nan, nan, nan,  ..., nan, nan, nan],
18:03:53         [nan, nan, nan,  ..., nan, nan, nan],
18:03:53         [nan, nan, nan,  ..., nan, nan, nan]], device='cuda:0')
18:03:53 analytical:tensor([[nan, nan, nan,  ..., nan, nan, nan],
18:03:53         [nan, nan, nan,  ..., nan, nan, nan],
18:03:53         [nan, nan, nan,  ..., nan, nan, nan],
18:03:53         ...,
18:03:53         [nan, nan, nan,  ..., nan, nan, nan],
18:03:53         [nan, nan, nan,  ..., nan, nan, nan],
18:03:53         [nan, nan, nan,  ..., nan, nan, nan]], device='cuda:0')
18:03:53 
18:03:53 
18:03:54 ----------------------------------------------------------------------
18:03:54 Ran 7794 tests in 2349.195s
18:03:54 
18:03:54 FAILED (errors=1, skipped=2435)
```

I always have to hack away a bit at the ONNX test failures, but it makes sense that ONNX doesn't know how to export the new symbol. Maybe a new function is needed here:

https://github.com/pytorch/pytorch/blob/c371542efc31b1abfe6f388042aa3ab0cef935f2/torch/onnx/symbolic_opset11.py#L442

**Review by IvanYashchuk — APPROVED** [review]
All my comments were addressed, so approved!

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 10 above)
Comment (antocuni) [reply aten/src/ATen/native/LinearAlgebra.cpp:63]: a9c99ef0cd

File: torch/testing/_internal/common_methods_invocations.py (hunk 11 above)
Comment (antocuni) [reply torch/testing/_internal/common_methods_invocations.py:28]: d962b375f4

File: torch/testing/_internal/common_methods_invocations.py (hunk 12 above)
Comment (antocuni) [reply torch/testing/_internal/common_methods_invocations.py:2828]: d962b375f4

**Comment by antocuni** [comment]
@mruberry 
I addressed the ONXX failure in a856c1a7fb: I also added an ONNX test specifically for `torch.linalg.det`, which looks like a good idea to me since it's the official one.

For ROCM, I just skipped test failing test in 8959ba96ca: `test_fn_grad` was already skipped, so it's not suprising that `test_fn_gradgrad` fails as well.

I also addressed the few nitpicks you mentioned.

**Comment by BowenBao** [comment]
@antocuni Thank you for updating the ONNX symbolic fn, the update looks good. For onnx test, I'd recommend also update the test here https://github.com/pytorch/pytorch/blob/26c1e2ee8300c679e0f8fafaf2b1acc61b54c806/test/onnx/test_pytorch_onnx_onnxruntime.py#L5268

**Comment by antocuni** [comment]
> @antocuni Thank you for updating the ONNX symbolic fn, the update looks good. For onnx test, I'd recommend also update the test here
> 
> https://github.com/pytorch/pytorch/blob/26c1e2ee8300c679e0f8fafaf2b1acc61b54c806/test/onnx/test_pytorch_onnx_onnxruntime.py#L5268

thank you for the suggestion @BowenBao , done in b39591fb50
